components/supervisor-api/notification.proto

```diff
@@ -11,60 +11,126 @@ import "google/api/annotations.proto";
+  // otherwise it blocks until the user has made their choice.
+  rpc Notify(NotifyRequest) returns (NotifyResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify"
+    };
+  }
+
+  // Subscribe to notifications. Typically called by the IDE.
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
+    option (google.api.http) = {
+      get : "/v1/notification/subscribe"
+    };
+  }
+
+  // Report a user's choice as a response to a notification. Typically called by
+  // the IDE.
+  rpc Respond(RespondRequest) returns (RespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/respond"
+    };
+  }
+
+  // SubscribeActive to inform supervisor about
+  // which is the latest client actively used by the user.
+  // Typically called by the IDE.
+  // Will response what client should do like `open` `preview`
+  rpc SubscribeActive(SubscribeActiveRequest)
+      returns (stream SubscribeActiveResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/subscribe-active"
+    };
+  }
+
+  rpc NotifyActive(NotifyActiveRequest) returns (NotifyActiveResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify-action"
+    };
+  }
+
+  rpc NotifyActiveRespond(NotifyActiveRespondRequest)
+      returns (NotifyActiveRespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify-action-respond"
+    };
+  }
 }
 
 message NotifyRequest {
-    enum Level {
-        ERROR = 0;
-        WARNING = 1;
-        INFO = 2;
-    }
-    Level level = 1;
-    string message = 2;
-    // if actions are empty, Notify will return immediately
-    repeated string actions = 3;
+  enum Level {
+    ERROR = 0;
+    WARNING = 1;
+    INFO = 2;
+  }
+  Level level = 1;
+  string message = 2;
+  // if actions are empty, Notify will return immediately
+  repeated string actions = 3;
 }
 
 message NotifyResponse {
-    // action chosen by the user or empty string if cancelled
-    string action = 1;
+  // action chosen by the user or empty string if cancelled
+  string action = 1;
 }
 
 message SubscribeRequest {}
 
 message SubscribeResponse {
-    uint64 requestId = 1;
-    NotifyRequest request = 2;
+  uint64 requestId = 1;
+  NotifyRequest request = 2;
 }
 
 message RespondRequest {
-    uint64 requestId = 1;
-    NotifyResponse response = 2;
+  uint64 requestId = 1;
+  NotifyResponse response = 2;
+}
+
+message RespondResponse {}
+
+message NotifyActiveRequest {
+  // open a file in editor
+  message OpenData {
+    repeated string urls = 1;
+    // wait until all opened files are closed
+    bool await = 2;
+  }
+
+  // ask editor to open a URL in its preview
+  message PreviewData {
+    string url = 1;
+    // open the URL in a new browser tab
+    bool external = 2;
+  }
+
+  oneof action_data {
+    OpenData open = 1;
+    PreviewData preview = 2;
+  }
+}
+
+message NotifyActiveResponse {
+  uint64 code = 1;
```

Comment:
it should be also oneof and correspond to kind of aciton in request.

Reply:
@akosyakov do you mean something like this?

```go
message NotifyActiveResponse {
  enum ActionType {
    OPEN = 0;
    PREVIEW = 1;
  }

  uint64 code = 1;
  string message = 2;
  ActionType action = 3;
}
```

cc @mustard-mh

Reply:
Not sure what to do with, if we do not care about result of action, it should be empty message.

If we do care about it, maybe something like this?

🅰️ 
```
message NotifyActiveResponse {
  message OpenData {
    uint64 code = 1;
    string message = 2;
  }

  message PreviewData {
    uint64 code = 1;
    string message = 2;
  }

  oneof data {
    OpenData open = 1;
    PreviewData preview = 2;
  }
}
```

🅱️ 
```
message NotifyActiveResponse {
  message OpenData {
    xxx
  }
  message PreviewData {
    xxx
  }

  uint64 code = 1;
  string message = 2;
  oneof data {
    OpenData open = 3;
    PreviewData preview = 4;
  }
}
```

I thought `code (status code)` and `message` is enough, so use one type for it

Reply:
Honestly for now I will leave `NotifyActiveResponse` just empty. I don't existing clients propagate anything back. For error reporting should not rather propagate using grpc errors?